distributed/worker_state_machine.py

```diff
@@ -231,20 +232,23 @@ class TaskState:
     #: True if the task is in memory or erred; False otherwise
     done: bool = False
 
+    _instances: ClassVar[weakref.WeakSet[TaskState]] = weakref.WeakSet()
+
     # Support for weakrefs to a class with __slots__
     __weakref__: Any = field(init=False)
 
+    def __post_init__(self):
+        TaskState._instances.add(self)
```

Comment:
This doesn't work on unpickle.

```suggestion
    def __new__(cls, *args, **kwargs):
        TaskState._instances.add(self)
        return object.__new__(cls)
```
\+ unit test for pickle/unpickle round trip

Reply:
We're not pickling TaskState objects anywhere, are we?

Reply:
We aren't today, but that's not an excuse to have it future-proofed. Namely it would make a lot of sense to pickle dump our new `WorkerState` class and everything it contains.

Reply:
Well, the "excuse" is scope creep.

Using __new__ as in 

```python
    def __new__(cls, *args, **kwargs):
        inst = object.__new__(cls)
        TaskState._instances.add(inst)
        return inst
```

does not work because we're defining the hash function as the hash of the key, i.e. we can only add fully initialized TaskState objects to the weakref.

Apart from this, we actually *can* (un-)pickle this class but will simply not add the instance to this weakset. For the only purpose of dumping this (like in our cluster dump) this is absolutely sufficient. 
At this point and for this functionality I'm not willing to reconsider the hash function